### DISCUSSION TITLE
Can i use aiocache in blacksheep

### DISCUSSION THREAD

**Discussion by yxlwfds** [Neoteroi/BlackSheep#220] [Q&A] [answered]
##### _Note: consider using [Discussions](https://github.com/Neoteroi/BlackSheep/discussions) to open a conversation about new features…_

**🚀 Feature Request**

**Comment by yxlwfds** [comment]
on router or function

**Comment by RobertoPrevato** [comment, marked as answer]
Hi @yxlwfds 
Sure, request handlers in BlackSheep support additional decorators, so they are compatible with any library. The order of decorators matters, but this is a common scenario in Python, it happens also with abstract properties: `@property` and `@abstractmethod`.

For example, this works, if you need this kind of functionality:

```python
import asyncio
from datetime import datetime

from aiocache import Cache, cached
from aiocache.serializers import PickleSerializer
from blacksheep.server.application import Application
from blacksheep.server.bindings import FromQuery
from blacksheep.server.responses import text

app = Application()


@app.route("/")
@cached(ttl=10, cache=Cache.MEMORY, key="key", serializer=PickleSerializer())
async def home(request):
    print("Sleeping for one second.....")
    await asyncio.sleep(1)
    return text(f"Hello, World! {datetime.utcnow().isoformat()}")

```

You can also register an instance of `Cache` as service, to have it automatically injected by dependency injection, see the function `other_example` below:

```python
import asyncio
from datetime import datetime

from aiocache import Cache, cached
from aiocache.backends.memory import SimpleMemoryCache
from aiocache.serializers import PickleSerializer
from blacksheep.server.application import Application
from blacksheep.server.bindings import FromQuery
from blacksheep.server.responses import text

app = Application()


# example using @cached decorator
@app.route("/")
@cached(ttl=10, cache=Cache.MEMORY, key="key", serializer=PickleSerializer())
async def home(request, x: FromQuery[int] = FromQuery(30)):
    print("Sleeping for one second.....")
    await asyncio.sleep(1)
    return text(f"Hello, World! {datetime.utcnow().isoformat()}")


# example using an injected service
app.services.add_instance(Cache(Cache.MEMORY))


@app.route("/other-example")
async def other_example(cache: SimpleMemoryCache):

    item = await cache.get("key")

    if item is None:
        await asyncio.sleep(1)  # simulate delay
        await cache.set("key", "Lorem ipsum")
        return text("The cached item was just set")
    else:
        return text(item)
```

In BlackSheep you cannot apply caching of responses to the `Router`, because the `Router` class has the single responsibility to match request URL's paths to request handlers, it is not concerned with the execution of those request handlers.
The equivalent would be to define a subclass of the `Application` class, override its `handle` method to cache the responses following the desired criteria.

Honestly I cannot think of any useful use-case for caching the responses globally this way (subclassing the `Application` to override `handle` and cache on all request-response cycles). I usually control the cache using `Cache-Control` response headers on the front-end layer, and cache in the business logic layer when I need caching like the one provided by `aiocache`.
